rpc/client.go

```diff
@@ -159,15 +161,15 @@ func (c *Client) connect(opts *retry.Options, context *Context) {
 		log.Infof("client %s connected", c.addr)
 		close(c.Ready)
 
-		// Launch periodic heartbeat.
-		go c.startHeartbeat()
-
 		return retry.Break, nil
-	})
-	if err != nil {
+	}); err != nil {
 		log.Errorf("client %s failed to connect: %v", c.addr, err)
 		c.Close()
 	}
+
+	// Launch periodic heartbeat.
+	c.runHeartbeat(context)
```

Comment:
if the single reason for passing the context in is access to the stopper, I would prefer passing in only the stopper for now (both into NewClient and runHeartbeat) since that's the pattern we have elsewhere, it shares less data and it's obvious from looking at it what it does and why.